### ISSUE TITLE
Request for deterministic support for reflection_pad2d_backward_cuda

### ISSUE THREAD

**Issue by rusvagzur** [pytorch/pytorch#98925]
### 🚀 The feature, motivation and pitch

Currently, the `reflection_pad2d_backward_cuda` operation used in neural network training cannot be computed deterministically on the GPU. This causes issues for users who have set `torch.use_deterministic_algorithms(True)` and require deterministic behavior in their training process.

I would like to request that deterministic support be added for `reflection_pad2d_backward_cuda` so that users can perform neural network training deterministically on the GPU. Thank you for your consideration.

**Comment by rusvagzur** [comment]
are there any updates?

**Comment by Yatagarasu50469** [comment]
It looks like you can force deterministic behavior with a high-level re-implementation; not optimized/optimal, but seems functional. Haven't tried it, but maybe wrapping an existing forward call of ReflectionPad2D in a manual module definition could work (might force torch to figure out the backwards path). Anyway, quick example for deterministic reflection padding to all sides of an input: 
```

#Deterministic reflection padding; modified from https://github.com/pytorch/vision/blob/main/torchvision/transforms/_functional_tensor.py#L322
class PadRef(nn.Module):
    def __init__(self, padding):
        super().__init__()
        self.register_buffer('indices_LeftTop', torch.tensor([i for i in range(padding, 0, -1)]))
        self.register_buffer('indices_RightBottom', torch.tensor([-(i + 2) for i in range(padding)]))
        
    def forward(self, data):
        x_indices = torch.cat([self.indices_LeftTop, torch.arange(data.shape[-1], device=data.device), self.indices_RightBottom])
        y_indices = torch.cat([self.indices_LeftTop, torch.arange(data.shape[-2], device=data.device), self.indices_RightBottom])
        return data[:, :, y_indices[:, None], x_indices[None, :]]
```

**Comment by albanD** [comment]
We would accept a PR implementing this for sure!

**Comment by ZelboK** [comment]
> We would accept a PR implementing this for sure!

I could give this a go. How impactful would this be?

**Comment by Yatagarasu50469** [comment]
Cannot speak as to the global impact, but it would be extremely useful for model ablations...
1) Can be used as the basis for a deterministic bilinear upsampling implementation (particularly applicable to U-Nets); where Bilinear: align_corners=False, half_pixel_centers=True = input with even dimensions + 2x nearest-exact upsampling + 3x3 binomial filter + reflection padding) 
2) Repeatable reduction of asymmetries/spatial-bias in CNN models ([Mind the Pad](https://arxiv.org/pdf/2010.02178)). 
3) Somewhat relatedly to 2, if there were a backward_cuda for deterministic reflection padding, then there would probably be more straightforward adaptions for deterministic symmetric and replication padding methods as well.

**Comment by ZelboK** [comment]
> Cannot speak as to the global impact, but it would be extremely useful for model ablations...
> 
> 1. Can be used as the basis for a deterministic bilinear upsampling implementation (particularly applicable to U-Nets); where Bilinear: align_corners=False, half_pixel_centers=True = input with even dimensions + 2x nearest-exact upsampling + 3x3 binomial filter + reflection padding)
> 2. Repeatable reduction of asymmetries/spatial-bias in CNN models ([Mind the Pad](https://arxiv.org/pdf/2010.02178)).
> 3. Somewhat relatedly to 2, if there were a backward_cuda for deterministic reflection padding, then there would probably be more straightforward adaptions for deterministic symmetric and replication padding methods as well.

I should have some time to give this a go. I have a solution in mind I just might need some assistance in validating the properties of the algorithm. Are you on the pytorch slack or discord anywhere?

**Comment by jmcoholich** [comment]
> It looks like you can force deterministic behavior with a high-level re-implementation; not optimized/optimal, but seems functional. Haven't tried it, but maybe wrapping an existing forward call of ReflectionPad2D in a manual module definition could work (might force torch to figure out the backwards path). Anyway, quick example for deterministic reflection padding to all sides of an input:
> 
> ```
> 
> #Deterministic reflection padding; modified from https://github.com/pytorch/vision/blob/main/torchvision/transforms/_functional_tensor.py#L322
> class PadRef(nn.Module):
>     def __init__(self, padding):
>         super().__init__()
>         self.register_buffer('indices_LeftTop', torch.tensor([i for i in range(padding, 0, -1)]))
>         self.register_buffer('indices_RightBottom', torch.tensor([-(i + 2) for i in range(padding)]))
>         
>     def forward(self, data):
>         x_indices = torch.cat([self.indices_LeftTop, torch.arange(data.shape[-1], device=data.device), self.indices_RightBottom])
>         y_indices = torch.cat([self.indices_LeftTop, torch.arange(data.shape[-2], device=data.device), self.indices_RightBottom])
>         return data[:, :, y_indices[:, None], x_indices[None, :]]
> ```

Here is a faster version (courtesy of ChatGPT o-1):

class PadRef(nn.Module):
    def __init__(self, padding):
        super().__init__()
        if isinstance(padding, list):
            assert all(x == padding[0] for x in padding), "All padding values must be equal."
            self.padding = padding[0]
        else:
            self.padding = padding

    def forward(self, data):
        p = self.padding
        if len(data.shape) != 4:
            raise ValueError("Only batched 4D tensors are supported (shape should be (batch, channels, height, width))")
        batch_size, channels, h, w = data.shape
        out_h = h + 2 * p
        out_w = w + 2 * p

        # Preallocate output tensor
        output = data.new_zeros((batch_size, channels, out_h, out_w))

        # Copy the central part
        output[:, :, p:p+h, p:p+w] = data

        if p > 0:
            # Reflect left and right edges
            # Left edge
            output[:, :, p:p+h, 0:p] = data[:, :, :, 1:p + 1].flip(-1)
            # Right edge
            output[:, :, p:p+h, p+w:] = data[:, :, :, -(p + 1): -1].flip(-1)

            # Reflect top and bottom edges
            # Top edge
            output[:, :, 0:p, :] = output[:, :, p + 1:p*2 + 1, :].flip(2)
            # Bottom edge
            output[:, :, p+h:, :] = output[:, :, h-1:h+p-1, :].flip(2)
        else:
            # If padding is zero, simply return the original data
            return data

        return output
